File: joonas/076.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    typedef pair<TreeNode*, char> pi;
    
    vector<pi> v;
    
    void dfs(TreeNode* p, char from) {
        if (p->left) dfs(p->left, -1);
        v.push_back(pi(p, from));
        if (p->right) dfs(p->right, 1);
    }
    
    bool sym(const pi& a, const pi& b) {
        return a.first->val == b.first->val && a.second + b.second == 0;
    }
    
    bool isSymmetric(TreeNode* root) {
        dfs(root, 0);
        int n = v.size();
        if (!(n & 1)) return false; // len(L-root-R) == odd
        for (int i=0; 2 * i < n - 1; ++i) {
            if (!sym(v[i], v[n-1-i])) return false;
        }
        return true;
    }
};

File: jlogkim/076.java
class Solution {
    public boolean checkSymmetric(TreeNode a, TreeNode b){
        //assume we should check nullness of node a and node b
        if(a==null&&b==null)return true;
        else if(a==null||b==null) return false;
        
        if(a.val!=b.val){return false;}
        if(checkSymmetric(a.left,b.right)&&checkSymmetric(a.right,b.left))
            return true;
        else 
            return false;
        
    }
    
    
    public boolean isSymmetric(TreeNode root) {
        return checkSymmetric(root.left,root.right);
    }
}